lib/phoenix/channel/server.ex

```diff
@@ -79,23 +88,55 @@ defmodule Phoenix.Channel.Server do
     :ok
   end
 
-  defp handle_result({:ok, socket}), do: {:noreply, socket}
-  defp handle_result({:leave, socket}), do: leave_and_stop(:normal, socket)
-  defp handle_result({:error, reason, socket}) do
+  defp handle_result({:reply, {status, response}, socket}, :handle_in) do
+    push socket, "phx_reply", %{status: to_string(status),
+                                ref: socket.ref,
+                                response: response}
+    {:noreply, socket}
+  end
+  defp handle_result({:reply, status, socket}, :handle_in) when is_atom(status) do
+    push socket, "phx_reply", %{status: to_string(status),
+                                ref: socket.ref,
+                                response: %{}}
+    {:noreply, socket}
+  end
+  defp handle_result({:reply, _, _socket}, _) do
+    raise """
+    Channel replies can only be sent from a `handle_in/3` callback.
+    Use `push/3` to send an out-of-bad message down the socket
+    """
+  end
+  defp handle_result({:noreply, socket}, _callback_type), do: {:noreply, socket}
+  defp handle_result({:leave, socket}, _callback_type), do: leave_and_stop(:normal, socket)
+  defp handle_result({:error, reason, socket}, _callback_type) do
     {:stop, {:error, reason}, socket}
   end
-  defp handle_result(result) do
+  defp handle_result(result, callback_type) do
     raise """
-      Expected callback to return `{:ok, socket} | {:error, reason, socket} || {:leave, socket}`,
-      got #{inspect result}
+    Expected `#{callback_type}` to return one of:
+
+        {:noreply, socket} |
+        {:reply, {status, response}, socket} |
+        {:reply, status, socket} |
+        {:error, reason, socket} |
+        {:leave, socket}
+
+    got #{inspect result}
     """
   end
 
   defp leave_and_stop(reason, socket) do
-    {:ok, socket} = socket.channel.leave(reason, socket)
-
     PubSub.unsubscribe(socket.pubsub_server, self, socket.topic)
```

Comment:
Remember we can't rely on this. Can we simply remove this line? Won't it automatically unsubscribe after it dies?